Create exportStream method

joshmossas opened the pull request:
## Change Summary
This pull request closes #87 

### Changes to `ApiCall.ts`

- `performRequest()` now accepts `responseType` as an optional parameter. That parameter is passed to Axios when making a request.
- `get()` now accepts `responseType` in it's optional parameters

### Changes to `Documents.ts`
- Added function called `exportStream`. It's functionally the same as `export()` the only difference is it now it passes `{ responseType: "stream" }` to the `this.apiCall.get()`.

### Additional Notes
[Axios doesn't support returning a stream in the browser](https://github.com/axios/axios/issues/479). So if this is called in the browser it'll just return a string like normal. Not sure if this needs to be documented or not. We could add a note to the JSDoc comment although I imagine most people wouldn't be using the export methods in the browser anyway since they require admin keys.

All tests are passing. I also did some basic tests in NodeJS and the Browser.

## PR Checklist
<!--- Put an `x` inside the box : -->
- [x] I have read and signed the [Contributor License Agreement](https://forms.gle/PZyiY5N2GDQU8GsV9).

joshmossas commented:
Also I'm not sure if I need to include the changes in `./lib` and `./dist`. Is this handled by a build system? Or should I included the newly outputted files in this PR?

jasonbosco commented:
Thank you @joshmossas! 🙏 

The `lib` and `dist` directories get built as part of the build/publishing process, so we can skip that as part of this PR. 

Could you also add a test for this new method similar to the one here: https://github.com/typesense/typesense-js/blob/b1d7ca1c930d09b95e55b44e4350d2b178d3aaf1/test/Typesense/Documents.spec.js#L376-L401

joshmossas commented:
Hey @jasonbosco 

I've added a test for .exportStream as requested. 

## Some Notes
- In order to test it properly I needed to create a temporary file so that I could create a readStream to return when mocking Axios. This temp file is written to `./test-files/exportStreamData.jsonl` and then deleted after the test completes. I've added `./test-files` to the gitignore. However I can use a different location if you prefer.
- I used the arrow function syntax for `directoryExists` and `getDataFromStream` because I was running into a conflict with Prettier and the Eslint rule `space-before-function-paren`. Prettier would format the function to remove the space between the function name and the space (Basically it would turn `function myFun ()` to `function myFunc()`) resulting in a linting error. Because of this I used the arrow syntax for these two methods to get around it. I only made note of this since the rest of the file uses the regular function syntax.

jasonbosco commented:
LGTM

jasonbosco commented:
Published this in `v1.0.3-3`

jasonbosco commented:
My bad. I just published `1.0.3-4` with the latest build.

joshmossas commented:
No worries. Thanks my dude.
